### PR TITLE
CU-25zu5uv Use Pablo in Dex-Router.

### PR DESCRIPTION
**Pull request by vivekvpandya**
## Issue
https://app.clickup.com/t/25zu5rm
https://app.clickup.com/t/25zu5uv

## Description
Use pablo pools in dex-router instead of curve-amm and uniswap pallets.
This will allow us to remove curve-amm and uniswap pallets.
This PR also improves checks on route parameter for update_route() to check transitivity of the route.

### CODE REVIEW

File: frame/composable-traits/src/dex.rs [hunk 1]
```diff
@@ -242,8 +242,9 @@ pub struct LiquidityBootstrappingPoolInfo<AccountId, AssetId, BlockNumber> {
 
 #[derive(Encode, Decode, MaxEncodedLen, TypeInfo, Clone, PartialEq, Eq, RuntimeDebug)]
 pub enum DexRouteNode<PoolId> {
-	Curve(PoolId),
-	Uniswap(PoolId),
+	// Curve(PoolId),
+	// Uniswap(PoolId),
```
Comment (Dom-Roth) [inline frame/composable-traits/src/dex.rs:246]: Do we want to keep these as comments?

**Review by Dom-Roth — APPROVED** [review]
LGTM!

File: frame/composable-traits/src/dex.rs (hunk 1 above)
Comment (vivekvpandya) [reply frame/composable-traits/src/dex.rs]: Fixed it. Thanks!

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-25zu5uv Add a check for non-transitive routes](https://app.clickup.com/t/25zu5uv)

File: frame/dex-router/src/lib.rs [hunk 2]
```diff
@@ -130,31 +126,47 @@ pub mod pallet {
...
+			ensure!(currency_pair.base == asset_pair.base, Error::<T>::UnexpectedNodeFound);
+			// check base asset of last node is quote asset of current node, i.e route must have
+			// transitivity
+			let mut last_base = asset_pair.quote;
+			for r in route.as_slice() {
+				let currency_pair = get_currency_pair(r)?;
+				ensure!(last_base == currency_pair.quote, Error::<T>::UnexpectedNodeFound);
+				last_base = currency_pair.base;
+			}
+			Ok(())
```
Comment (hussein-aitlahcen) [inline frame/dex-router/src/lib.rs:158]: You could make this a lot easier with a try_fold, starting with an error and then shifting to OK if transitivity is respected

**Comment by hussein-aitlahcen** [comment]
@vivekvpandya not sure why we did not implemented the Amm trait on dex router here?

**Comment by vivekvpandya** [comment]
> 

Sorry I don't understand why it is required to implement Amm trait for dex-router?
Do we want to use dex-router as dex it self?

**Comment by hussein-aitlahcen** [comment]
> > 
> 
> Sorry I don't understand why it is required to implement Amm trait for dex-router? Do we want to use dex-router as dex it self?

Yes

**Comment by vivekvpandya** [comment]
> I mean the parameter like min_receive must be used. So the interface will be really similar, we can probably implement Amm with PoolId=CurrencyPair ? Not sure but would be nice if possible

I prefer to do it in different PR. My motivation for this PR is to remove Uniswap and CurveAmm pallet uses from dex-router so that if we want to delete them we can. So if its okay let's complete this PR and open a separate click-up ticket for implement Amm for dex-router.

File: frame/dex-router/src/lib.rs [hunk 3]
```diff
@@ -109,52 +105,66 @@ pub mod pallet {
...
+			old_route: Vec<T::PoolId>,
+			updated_route: Vec<T::PoolId>,
 		},
 	}
 
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {}
 
 	impl<T: Config> Pallet<T> {
+		fn check_route(
```
Comment (hussein-aitlahcen) [inline frame/dex-router/src/lib.rs:129]: Can we have tests checking the failure case?

Comment (vivekvpandya) [reply frame/dex-router/src/lib.rs:129]: update_route() test covers all possible errors.
https://github.com/ComposableFi/composable/pull/899/files#diff-183beb853fd7716a87bad38cf794b5a91acdf4cf98dd86d52e2a499760ad5abfR148

**Review by hussein-aitlahcen — APPROVED** [review]

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-276hz1f Polkadot.js Docs](https://app.clickup.com/t/276hz1f)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-276hz1f Polkadot.js Docs](https://app.clickup.com/t/276hz1f)

**Review by dzmitry-lahoda — APPROVED** [review]

**Pull request merged**